c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -457,9 +461,92 @@ impl<'a, 'tcx> HirTyVisitor<'a, 'tcx> {
+        let mut first_generic_const_span: Option<Span> = None;
+
+        for (predicate, predicate_span) in predicates.predicates {
+            if matches!(
+                predicate.kind().skip_binder(),
+                PredicateKind::RegionOutlives(OutlivesPredicate(a, b)) if a != b
+            ) && (last_lifetime_span.is_none()
+                || last_lifetime_span.unwrap().hi() < predicate_span.hi())
+            {
+                last_lifetime_span = Some(*predicate_span)
+            }
+        }
+        for param in generics.params {
+            match param.kind {
+                GenericParamKind::Lifetime { .. } => {
+                    if last_lifetime_span.is_none()
+                        || last_lifetime_span.unwrap().hi() < param.span.hi()
+                    {
+                        last_lifetime_span = Some(param.span)
+                    }
+                }
+                GenericParamKind::Type { .. } => {
+                    if first_generic_type_span.is_none()
+                        || first_generic_type_span.unwrap().lo() < param.span.lo()
+                    {
+                        first_generic_type_span = Some(param.span)
+                    }
+                }
+                GenericParamKind::Const { .. } => {
+                    if first_generic_const_span.is_none()
+                        || first_generic_const_span.unwrap().lo() < param.span.lo()
+                    {
+                        first_generic_const_span = Some(param.span)
+                    }
+                }
+            }
+        }
+
+        let generics_spans = (
+            last_lifetime_span,
+            first_generic_type_span,
+            first_generic_const_span,
+        );
+
+        let hypothetical_origin_params: Vec<_> = origin_params
+            .filter_map(|p| {
+                if let OriginParam::Hypothetical(_) = p {
+                    Some(format!("{:?}", p))
+                } else {
+                    None
+                }
+            })
+            .collect();
+        let hypothetical_origin_string = hypothetical_origin_params.join(",");
+
+        if !hypothetical_origin_string.is_empty() {
+            let (hypothetical_origin_span, format_string) = match generics_spans {
+                (Some(last_lifetime_span), ..) => (
+                    last_lifetime_span.shrink_to_hi(),
+                    format!(",{}", hypothetical_origin_string),
+                ),
+                (_, Some(first_generics_span), _) | (_, _, Some(first_generics_span)) => (
+                    first_generics_span.shrink_to_lo(),
+                    format!("{},", hypothetical_origin_string),
+                ),
+                _ if generics.params.is_empty() => {
+                    (generics.span, format!("<{}>", hypothetical_origin_string))
+                }
+                _ => (generics.span, format!("{}", hypothetical_origin_string)),
```

Comment:
I think the `if` condition here is wrong.  If `last_lifetime_span`, `first_ty_span`, and `first_const_span` are all `None`, then we already know the parameter list is empty.  What we actually need to distinguish here are `fn foo()` vs `fn foo<>()`.  Both will have an empty `params` list, so my guess is the only way to distinguish these is to check whether `generics.span` is empty or not (`lo() == hi()`, or maybe there's a `Span` method to check for this specifically).